core-backend/sandbox/src/funcs.rs

```diff
@@ -278,23 +277,25 @@ where
 
     /// Fallible `gr_read` syscall.
     pub fn read(ctx: &mut Runtime<E>, args: &[Value]) -> SyscallOutput {
-        sys_trace!(target: "syscall::gear", "read, args = {}", args_to_str(args));
+        sys_trace!(target: "syscalls", "read, args = {}", args_to_str(args));
 
         let (at, len, buffer_ptr, err_len_ptr) = args.iter().read_4()?;
 
         ctx.run_fallible::<_, _, LengthBytes>(err_len_ptr, RuntimeCosts::Read, |ctx| {
-            let buffer = ctx.ext.read(at, len)?;
+            // Here `ctx.ext` is const borrowed, so we cannot use `ctx` mut methods.
+            let (buffer, mut gas_left) = ctx.ext.read(at, len)?;
 
             let write_buffer = ctx.memory_manager.register_write(buffer_ptr, len);
             ctx.memory_manager
-                .write(&mut ctx.memory, write_buffer, buffer)
-                .map_err(Into::into)
+                .write(&mut ctx.memory, write_buffer, buffer, &mut gas_left)?;
```

Comment:
shouldn't we set gas left before quitting closure?

Reply:
add issue to solve the problem of returning gas left from read  fn by design

Reply:
add TODO in env.rs